pkg/cmd/roachprod-microbench/cluster/execute.go

```diff
@@ -34,47 +34,93 @@ type RemoteCommand struct {
 // A Duration of -1 indicates that the command was cancelled.
 type RemoteResponse struct {
 	RemoteCommand
-	Stdout     string
-	Stderr     string
-	Err        error
-	ExitStatus int
-	Duration   time.Duration
+	Stdout        string
+	Stderr        string
+	Err           error
+	ExitStatus    int
+	Duration      time.Duration
+	commandStatus CommandStatus
 }
 
+type CommandStatus int8
+
+const (
+	Completed CommandStatus = iota
+	Cancelled
+	Terminated
+)
+
+var ErrAllNodesTerminated = errors.New("all nodes terminated")
+
+type RemoteExecutionFunc func(
+	ctx context.Context,
+	l *logger.Logger,
+	clusterName, SSHOptions, processTag string,
+	secure bool,
+	cmdArray []string,
+	options install.RunOptions,
+) ([]install.RunResultDetails, error)
+
 func remoteWorker(
 	ctx context.Context,
 	log *logger.Logger,
+	execFunc RemoteExecutionFunc,
 	clusterNode string,
-	receive chan []RemoteCommand,
-	response chan RemoteResponse,
+	workChan chan []RemoteCommand,
+	responseChan chan RemoteResponse,
 ) {
 	for {
-		commands := <-receive
+		commands := <-workChan
 		if commands == nil {
 			return
 		}
 		for index, command := range commands {
 			if errors.Is(ctx.Err(), context.Canceled) {
 				for _, cancelCommand := range commands[index:] {
-					response <- RemoteResponse{cancelCommand, "", "", nil, -1, -1}
+					responseChan <- RemoteResponse{RemoteCommand: cancelCommand, commandStatus: Cancelled}
 				}
 				break
 			}
 			start := timeutil.Now()
-			runResult, err := roachprod.RunWithDetails(
+			runResult, err := execFunc(
 				context.Background(), log, clusterNode, "" /* SSHOptions */, "", /* processTag */
-				false /* secure */, command.Args, install.DefaultRunOptions(),
+				false /* secure */, command.Args, install.DefaultRunOptions().WithRetryDisabled(),
 			)
 			duration := timeutil.Since(start)
+
+			// Reschedule on another worker when a `roachprod` or SSH error occurs. A
+			// `roachprod` error is when `result.Err == nil`.
+			if len(runResult) > 0 {
+				err = errors.CombineErrors(err, runResult[0].Err)
+			}
+			if err != nil && (runResult[0].Err == nil || errors.Is(err, rperrors.ErrSSH255)) {
```

Comment:
Do you still need this complicated predicate now that we call `CombineErrors`? I think just `errors.Is(err, rperrors.ErrSSH255)` should suffice.